### ISSUE TITLE
Devmode thread DX test shows 404 page for Jetty

### ISSUE THREAD

**Issue by mshabarov** [vaadin/flow#7766]
Jetty error page with a 404 code is shown when the https://github.com/vaadin/bookstore-example project is built and run under `dx-2` branch (devmode thread feature branch). This issue is not reproduced under `dx-1` branch (without devmode thread feature). See attached screenshot and log file.

Steps to reproduce:

1. Run git clone https://github.com/vaadin/bookstore-example 
2. Run git checkout dx-2
3. Run the project from IDE (jetty:run)
4. Open app in Browser when `[INFO] Started Jetty Server` message appears

And for further attempts:

1. Delete node_modules folder
2. Run mvn clean
3. Open app in Browser when `[INFO] Started Jetty Server` message appears

![bookstore-dx-2-404](https://user-images.githubusercontent.com/61410877/76152794-8342a800-60cc-11ea-88b3-e43fba721ee0.jpg)
[DX-2-log.txt](https://github.com/vaadin/flow/files/4302528/DX-2-log.txt)

**Comment by mshabarov** [comment]
UPDATE:

I noticed the restart of Jetty context in the middle of app starting on top of `dx-2` branch:

```
[INFO] Started ServerConnector@35590f53{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
[INFO] Started @25613ms
[INFO] Started Jetty Server
.....
[INFO] restarting o.e.j.m.p.JettyWebAppContext@7e3cea61{/,...........}
[INFO] Stopped o.e.j.m.p.JettyWebAppContext@7e3cea61{.............}
.....
[INFO] Restart completed at Sat Mar 07 23:54:46 MSK 2020
```

There are no such a restarts in the dx-1 runs, so i believe this is a main suspect of the problem with 404 page. Gonna sort out what is causing this restart...

I also tried with latest version of Jetty (`9.4.27.v20200227`). Although, i didn't see the exception:

`Exception in thread "PathWatcher@127350d8" java.lang.NullPointerException`

but the issue with 404 is being reproduced.

**Comment by mshabarov** [comment]
Jetty Context restart is caused by `jetty-maven-plugin` file scanner.

Property scanIntervalSeconds is set to 2 in `pom.xml` of `bookstore-example`, which means the scanner checks for resources changes each 2 seconds, see [jetty docs](https://www.eclipse.org/jetty/documentation/9.4.x/jetty-maven-plugin.html)

Devmode Thread apparently updates some files, such that for example `generated-flow-imports.js` and this causes hot restart.

Setting this property to 0 seconds disables the scanner and neither restart is being observed nor the 404 page. I need to find why this property had been set initially, but the application is started fine even with 0 seconds interval.

So, this can be considered as a workaround for `Jetty`, but the other servers (i.e. `WebLogic`) have apparently a similar tools for scanning.

**Comment by mshabarov** [comment]
As it turned out, Jetty restart is being caused by `/node_modules` dir update.
This issue is proposed to be fixed by excluding `/node_modules` from Jetty scanner list.

**Comment by mshabarov** [comment]
Not only `node_modules` folder update triggers Jetty restart. Actually, Jetty scanner watches to
all the folder in project root (`frontend`, `target`, `src`, ...), however it is not mentioned in [docs](https://www.eclipse.org/jetty/documentation/current/jetty-maven-scanning.html). 

Moreover, there is one more directory Jetty is triggered on while restart - `/target/classes/META-INF/VAADIN/build`. Devmode thread creates `build` folder within `VAADIN` while running.

This scan list is mentioned in documentation:

- pom.xml
- `dependencies`
- `classesDirectory`
- `testClassesDirectory`
- `webXml` or `webAppSourceDirectory`/WEB-INF/web.xml
- `jettyEnvXml` or `webAppSourceDirectory`/WEB-INF/jetty-web.xml
- `webAppSourceDirectory`/WEB-INF/jetty-web.xml
- `scanTargets`
- `scanTargetPatterns`
- any `defaultsDescriptor` for the webapp
- any `overrideDescriptor` for the webapp
- any `dependencies` that are wars or zips

I tried to exclude `node_modules` and `VAADIN/build`directories from Jetty scan configuration like this:
```
<configuration>
        <scanTargetPatterns>
          <scanTargetPattern>
            <directory>node_modules</directory>
            <excludes>
              <exclude>**</exclude>
            </excludes>
          </scanTargetPattern>
        </scanTargetPatterns>

        <scanClassesPattern>
                  <excludes>
                     <exclude>META-INF/VAADIN/**</exclude>
                  </excludes>
         </scanClassesPattern>
</configuration>
```
        
but it didn't work and moreover, even if it does, it will bring an extra complexity to user-facing pom.xml, which is definitely decrease the DX.

**Comment by mshabarov** [comment]
UPDATE:

1. Extra tests have been performed on versions `9.4.22.v20191022` (old) and `9.4.27.v20200227` (latest) of `jetty-maven-plugin`.
2. Jetty source code analysis and tests show different behaviour in Jetty hot-swap scanner between those two versions:
- `latest` version does not monitor directories/files in the root of project (excluding `pom.xml` which is monitored by default), but the `old` version does.
- `old` version occasionally shows `NPE` exception in `PathWatcher` during Jetty startup, which interrupts hot-swap scanning. This exception has a stochastic nature and has been already observed in the past. I guess it may be caused by `webpack` and `Jetty scanner `concurrent runs.
- the same scanner configuration works differently on those two versions.

3. Those differences mentioned in 2. leads to the following:
- `latest` version triggers restart only for `VAADIN/build` folder during restart.
- `old` version triggers restart for **both** `VAADIN/build` and `node_modules` folders.
- `old` version occasionally fails to start the scanner because of NPE, so this disables hot-swap at all.

4. The optimal scanner configuration I found so far for `latest` version is the following:
```
    <scanClassesPattern>
        <excludes>
            <exclude>${project.build.outputDirectory}/META-INF/VAADIN/**</exclude>
            <exclude>${project.build.outputDirectory}/META-INF/VAADIN</exclude>
        </excludes>
    </scanClassesPattern>
```
It simply excludes `VAADIN` and `VAADIN/build` folders from scan list. 
No need to exclude `node_modules` because it is not monitored.
Developer is free to add a custom exclusions/inclusions if he wants.

5. This is the solution for `old` version:
 ```
<scanClassesPattern>
         <excludes>
             <exclude>META-INF/VAADIN</exclude>
         </excludes>
         <includes>
             <include>**/*.class</include>
             <include>**/*.properties</include>
         </includes>
     </scanClassesPattern>

     <scanTargetPatterns>
         <scanTargetPattern>
             <directory>.</directory>
             <excludes>
                 <exclude>*/**</exclude>
             </excludes>
         </scanTargetPattern>
     </scanTargetPatterns>
```
It has a couple of disadvantages: it's tricky, non-intuitive, redundant, and moreover it blocks the developer to add any custom folders to scan, because there is already a high-priority exclude statement skipping all the top-level folders in the root and its content.

6. So, as for me, we should't use 5., but should use the `latest` version and config from 4. instead.

**Comment by mshabarov** [comment]
Ticket is closed because there is a new ticket opened #7922 with further actions.
